[CIR] Upstream initial attribute support

dkolsen-pgi opened the pull request:
Upstream several ClangIR-specific MLIR attributes, in particular attributes for integer, floating-point, and null pointer constants. These are the first ClangIR attributes to be upstreamed, so infrastructure changes are included, such as the table-gen file `CIRAttrs.td`.

Attributes can be used as the initial values for global variables.  The existing automated test global-var-simple.cpp includes initial values for some of the global variables in the test.

dkolsen-pgi committed:
[CIR] Upstream initial attribute support

Upstream several ClangIR-specific MLIR attributes, in particular
attributes for integer, floating-point, and null pointer constants.
These are the first ClangIR attributes to be upstreamed, so
infrastructure changes are included, such as the table-gen file
`CIRAttrs.td`.

Attributes can be used as the initial values for global variables.  The
existing automated test global-var-simple.cpp includes initial values
for some of the global variables in the test.

erichkeane commented on clang/include/clang/CIR/Dialect/Builder/CIRBaseBuilder.h:
```diff
@@ -26,6 +30,13 @@ class CIRBaseBuilderTy : public mlir::OpBuilder {
     return getPointerTo(cir::VoidType::get(getContext()));
   }
+
+  mlir::TypedAttr getConstPtrAttr(mlir::Type t, int64_t v) {
```
Could we do a better variable name than `v`?  I presume it means `value`?  But that is really only something I can guess at.

dkolsen-pgi replied:
Done.

I am still calibrating when to use the same names that are in the ClangIR incubator and when to choose better names.

erichkeane commented on clang/lib/CIR/Dialect/IR/CIRAttrs.cpp:
```diff
@@ -21,18 +39,160 @@ using namespace cir;
+// ConstPtrAttr definitions
+//===----------------------------------------------------------------------===//
+
+// TODO(CIR): Consider encoding the null value differently and use conditional
```
Agreed with this TODO.  The textual representation of 'null' is a little fragile for me.

erichkeane replied:
^^ Nothing to do in this patch, just highlighting this TODO for posterity.

erichkeane commented on clang/lib/CIR/Dialect/IR/CIRAttrs.cpp:
```diff
@@ -21,18 +39,160 @@ using namespace cir;
+  if (type.isSigned()) {
+    int64_t value;
+    if (parser.parseInteger(value))
+      parser.emitError(parser.getCurrentLocation(), "expected integer value");
```
Do these need to `return`?  It seems that continuing with a garbage `value` here could be surprising?

dkolsen-pgi replied:
I don't want to add a return here because the code should fall through and try to parse the closing `>` character.  That increases the chance that the parser will recover from an invalid integer value and not get completely lost.  But I did rearrange the code some so that an uninitialized garbage value is never returned.

erichkeane replied:
Makes sense, this LGTM.

erichkeane reviewed (commented):
Few small things, else LGTM.  I'm away until next Thursday, so the others can approved if they are around.

dkolsen-pgi committed:
[CIR] Upstream attributes followup

Better parameter names for `CIRBaseBuilderTy::getConstPtrAttr`

Slightly better error handling for `IntAttr::parse`

erichkeane reviewed (approved):
Looks ok to me.
